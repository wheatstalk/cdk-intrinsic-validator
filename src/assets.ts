import * as path from 'path';

export const LAMBDA_ASSET_DIR = path.join(__dirname, '..', 'lambda');