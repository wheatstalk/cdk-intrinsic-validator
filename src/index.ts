export * from './intrinsic-validator';
export { LAMBDA_ASSET_DIR } from './assets';