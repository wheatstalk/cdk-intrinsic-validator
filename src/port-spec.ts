export function compilePortSpec(portSpec: string): (x: number) => boolean {
  const portsArray = portSpec.split(',')
    .flatMap(parseToPortList);

  const ports = new Set<number>(portsArray);

  return (x) => ports.has(x);
}

/**
 * Parses a port spec conjunct.
 *
 * @param portSpec
 * @returns the expansion of all values
 *
 * @example "200" -> [200], "500-599" => [500, 501, ..., 599]
 */
function parseToPortList(portSpec: string): number[] {
  if (IS_PORT.test(portSpec)) {
    return [parseInt(portSpec)];
  }

  if (IS_RANGE.test(portSpec)) {
    const [begin, end] = portSpec.split('-', 2);
    if (typeof begin === 'string' && typeof end === 'string') {
      const beginN = parseInt(begin);
      const endN = parseInt(end);
      return Array.from(Array(endN - beginN + 1).keys()).map(x => x + beginN);
    }
  }

  throw new Error(`Invalid port spec: ${portSpec}`);
}

const IS_PORT = /^\d+$/;
const IS_RANGE = /^\d+-\d+$/;