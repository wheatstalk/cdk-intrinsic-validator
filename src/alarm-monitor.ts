import * as cdk from 'aws-cdk-lib';
import * as cw from 'aws-cdk-lib/aws-cloudwatch';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as sfn from 'aws-cdk-lib/aws-stepfunctions';
import * as sfn_tasks from 'aws-cdk-lib/aws-stepfunctions-tasks';
import { Construct } from 'constructs';
import { LAMBDA_ASSET_DIR } from './assets';

/**
 * A CloudWatch alarm monitor.
 */
export interface IAlarmMonitor {
  /**
   * The Step Functions state machine.
   */
  readonly stateMachine: sfn.IStateMachine;

  /**
   * Create input for a state machine execution.
   */
  createTaskInput(options: AlarmMonitorTaskInputOptions): sfn.TaskInput;
}

/**
 * Task input options.
 */
export interface AlarmMonitorTaskInputOptions {
  /** The alarm to monitor */
  readonly alarm: cw.IAlarm;
  /** The maximum duration to monitor */
  readonly duration: cdk.Duration;
}

/**
 * Provides a State Machine to monitor CloudWatch Alarms.
 */
export class AlarmMonitor extends Construct implements IAlarmMonitor {
  readonly stateMachine: sfn.IStateMachine;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    const bugState = new sfn.Fail(this, 'BugState');

    const checkAlarmStatusFunction = new lambda.Function(this, 'CheckAlarmStatusFunction', {
      runtime: lambda.Runtime.NODEJS_14_X,
      handler: 'check-alarm-status.handler',
      code: lambda.Code.fromAsset(LAMBDA_ASSET_DIR),
      logRetention: logs.RetentionDays.ONE_MONTH,
      initialPolicy: [
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: ['cloudwatch:DescribeAlarms'],
          resources: ['*'],
        }),
      ],
    });

    const checkAlarmStatus = new sfn_tasks.LambdaInvoke(this, 'CheckAlarmStatus', {
      integrationPattern: sfn.IntegrationPattern.REQUEST_RESPONSE,
      lambdaFunction: checkAlarmStatusFunction,
      payload: sfn.TaskInput.fromObject({
        'AlarmName.$': '$.AlarmName',
        'MonitoringDurationSeconds.$': '$.MonitoringDurationSeconds',
        'StartTime.$': '$$.Execution.StartTime',
      }),
      resultPath: '$.CheckAlarmStatus',
      resultSelector: {
        'Status.$': '$.Payload.Status',
      },
    });

    checkAlarmStatus.addCatch(bugState);
    checkAlarmStatus.addRetry({ maxAttempts: 3 });

    const alarmStatusChoice = new sfn.Choice(this, 'AlarmStatusChoice');
    const delayForNextCheckStatus = new sfn.Wait(this, 'DelayThenNextCheckStatus', {
      time: sfn.WaitTime.duration(cdk.Duration.seconds(30)),
    });
    const failBecauseAlarming = new sfn.Fail(this, 'FailBecauseAlarming');
    const noAlarm = new sfn.Succeed(this, 'NoAlarm');

    const definition = sfn.Chain.start(checkAlarmStatus)
      .next(alarmStatusChoice
        .when(
          sfn.Condition.stringEquals('$.CheckAlarmStatus.Status', 'ALARM'),
          failBecauseAlarming,
        )
        .when(
          sfn.Condition.stringEquals('$.CheckAlarmStatus.Status', 'MONITORING'),
          delayForNextCheckStatus
            .next(checkAlarmStatus),
        )
        .when(
          sfn.Condition.stringEquals('$.CheckAlarmStatus.Status', 'NO_ALARM'),
          noAlarm,
        )
        .otherwise(bugState),
      );

    this.stateMachine = new sfn.StateMachine(this, 'StateMachine', {
      definition: definition,
    });
  }

  createTaskInput(options: AlarmMonitorTaskInputOptions): sfn.TaskInput {
    return sfn.TaskInput.fromObject({
      AlarmName: options.alarm.alarmName,
      MonitoringDurationSeconds: options.duration.toSeconds(),
    });
  }
}

/**
 * Creates or re-uses a singleton alarm monitor.
 */
export class SingletonAlarmMonitor extends Construct implements IAlarmMonitor {
  private readonly alarmMonitor: IAlarmMonitor;

  public get stateMachine(): sfn.IStateMachine {
    return this.alarmMonitor.stateMachine;
  }

  constructor(scope: Construct, id: string) {
    super(scope, id);

    const stack = cdk.Stack.of(this);
    this.alarmMonitor = stack.node.tryFindChild('SingletonAlarmMonitor') as AlarmMonitor
      ?? new AlarmMonitor(stack, 'SingletonAlarmMonitor');
  }

  createTaskInput(options: AlarmMonitorTaskInputOptions): sfn.TaskInput {
    return this.alarmMonitor.createTaskInput(options);
  }
}