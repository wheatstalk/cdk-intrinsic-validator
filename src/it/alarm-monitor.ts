import * as cdk from '@aws-cdk/core';
import { AlarmMonitor } from '../alarm-monitor';
import { TestAlarms } from './test-alarms';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'AlarmMonitor', {
  env: {
    account: process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEFAULT_REGION,
  },
});

new AlarmMonitor(stack, 'AlarmMonitor');
new TestAlarms(stack, 'TestAlarms');

// Submit the JSON for the Test Alarms to the state machine to test this thing.