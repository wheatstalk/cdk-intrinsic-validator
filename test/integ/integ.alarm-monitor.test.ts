import { SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { DisableRandomnessContextKey } from '../../src';
import { IntegAlarmMonitor } from './integ.alarm-monitor';

test('snapshot', () => {
  const app = new cdk.App({
    context: {
      [DisableRandomnessContextKey]: true,
    },
  });

  const stack = new IntegAlarmMonitor(app);
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});