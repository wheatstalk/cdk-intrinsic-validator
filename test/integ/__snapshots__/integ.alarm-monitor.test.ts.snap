// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
Object {
  "Mappings": Object {
    "ServiceprincipalMap": Object {
      "af-south-1": Object {
        "states": "states.af-south-1.amazonaws.com",
      },
      "ap-east-1": Object {
        "states": "states.ap-east-1.amazonaws.com",
      },
      "ap-northeast-1": Object {
        "states": "states.ap-northeast-1.amazonaws.com",
      },
      "ap-northeast-2": Object {
        "states": "states.ap-northeast-2.amazonaws.com",
      },
      "ap-northeast-3": Object {
        "states": "states.ap-northeast-3.amazonaws.com",
      },
      "ap-south-1": Object {
        "states": "states.ap-south-1.amazonaws.com",
      },
      "ap-south-2": Object {
        "states": "states.ap-south-2.amazonaws.com",
      },
      "ap-southeast-1": Object {
        "states": "states.ap-southeast-1.amazonaws.com",
      },
      "ap-southeast-2": Object {
        "states": "states.ap-southeast-2.amazonaws.com",
      },
      "ap-southeast-3": Object {
        "states": "states.ap-southeast-3.amazonaws.com",
      },
      "ap-southeast-4": Object {
        "states": "states.ap-southeast-4.amazonaws.com",
      },
      "ca-central-1": Object {
        "states": "states.ca-central-1.amazonaws.com",
      },
      "cn-north-1": Object {
        "states": "states.cn-north-1.amazonaws.com",
      },
      "cn-northwest-1": Object {
        "states": "states.cn-northwest-1.amazonaws.com",
      },
      "eu-central-1": Object {
        "states": "states.eu-central-1.amazonaws.com",
      },
      "eu-central-2": Object {
        "states": "states.eu-central-2.amazonaws.com",
      },
      "eu-north-1": Object {
        "states": "states.eu-north-1.amazonaws.com",
      },
      "eu-south-1": Object {
        "states": "states.eu-south-1.amazonaws.com",
      },
      "eu-south-2": Object {
        "states": "states.eu-south-2.amazonaws.com",
      },
      "eu-west-1": Object {
        "states": "states.eu-west-1.amazonaws.com",
      },
      "eu-west-2": Object {
        "states": "states.eu-west-2.amazonaws.com",
      },
      "eu-west-3": Object {
        "states": "states.eu-west-3.amazonaws.com",
      },
      "il-central-1": Object {
        "states": "states.il-central-1.amazonaws.com",
      },
      "me-central-1": Object {
        "states": "states.me-central-1.amazonaws.com",
      },
      "me-south-1": Object {
        "states": "states.me-south-1.amazonaws.com",
      },
      "sa-east-1": Object {
        "states": "states.sa-east-1.amazonaws.com",
      },
      "us-east-1": Object {
        "states": "states.us-east-1.amazonaws.com",
      },
      "us-east-2": Object {
        "states": "states.us-east-2.amazonaws.com",
      },
      "us-gov-east-1": Object {
        "states": "states.us-gov-east-1.amazonaws.com",
      },
      "us-gov-west-1": Object {
        "states": "states.us-gov-west-1.amazonaws.com",
      },
      "us-iso-east-1": Object {
        "states": "states.amazonaws.com",
      },
      "us-iso-west-1": Object {
        "states": "states.amazonaws.com",
      },
      "us-isob-east-1": Object {
        "states": "states.amazonaws.com",
      },
      "us-west-1": Object {
        "states": "states.us-west-1.amazonaws.com",
      },
      "us-west-2": Object {
        "states": "states.us-west-2.amazonaws.com",
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AlarmMonitorCheckAlarmStatusFunction38993418": Object {
      "DependsOn": Array [
        "AlarmMonitorCheckAlarmStatusFunctionServiceRoleDefaultPolicyF834009A",
        "AlarmMonitorCheckAlarmStatusFunctionServiceRole0FE14751",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "309ef33c6575fb63e8069c2f6f1dbb7d5234bbd8d9be7c666ea023ebdf1a6dc3.zip",
        },
        "Handler": "check-alarm-status.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AlarmMonitorCheckAlarmStatusFunctionServiceRole0FE14751",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "AlarmMonitorCheckAlarmStatusFunctionLogRetention1E4E68E2": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "AlarmMonitorCheckAlarmStatusFunction38993418",
              },
            ],
          ],
        },
        "RetentionInDays": 30,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "AlarmMonitorCheckAlarmStatusFunctionServiceRole0FE14751": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AlarmMonitorCheckAlarmStatusFunctionServiceRoleDefaultPolicyF834009A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "cloudwatch:DescribeAlarms",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AlarmMonitorCheckAlarmStatusFunctionServiceRoleDefaultPolicyF834009A",
        "Roles": Array [
          Object {
            "Ref": "AlarmMonitorCheckAlarmStatusFunctionServiceRole0FE14751",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "AlarmMonitorStateMachineC704D11C": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "AlarmMonitorStateMachineRoleDefaultPolicy7C3EF13E",
        "AlarmMonitorStateMachineRole19D08AAF",
      ],
      "Properties": Object {
        "DefinitionString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"StartAt\\":\\"CheckAlarmStatus\\",\\"States\\":{\\"CheckAlarmStatus\\":{\\"Next\\":\\"AlarmStatusChoice\\",\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ClientExecutionTimeoutException\\",\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2},{\\"ErrorEquals\\":[\\"States.ALL\\"],\\"MaxAttempts\\":3}],\\"Catch\\":[{\\"ErrorEquals\\":[\\"States.ALL\\"],\\"Next\\":\\"BugState\\"}],\\"Type\\":\\"Task\\",\\"ResultPath\\":\\"$.CheckAlarmStatus\\",\\"ResultSelector\\":{\\"Status.$\\":\\"$.Payload.Status\\"},\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "AlarmMonitorCheckAlarmStatusFunction38993418",
                  "Arn",
                ],
              },
              "\\",\\"Payload\\":{\\"AlarmName.$\\":\\"$.AlarmName\\",\\"MonitoringDurationSeconds.$\\":\\"$.MonitoringDurationSeconds\\",\\"StartTime.$\\":\\"$$.Execution.StartTime\\"}}},\\"DelayThenNextCheckStatus\\":{\\"Type\\":\\"Wait\\",\\"Seconds\\":30,\\"Next\\":\\"CheckAlarmStatus\\"},\\"AlarmStatusChoice\\":{\\"Type\\":\\"Choice\\",\\"Choices\\":[{\\"Variable\\":\\"$.CheckAlarmStatus.Status\\",\\"StringEquals\\":\\"ALARM\\",\\"Next\\":\\"FailBecauseAlarming\\"},{\\"Variable\\":\\"$.CheckAlarmStatus.Status\\",\\"StringEquals\\":\\"MONITORING\\",\\"Next\\":\\"DelayThenNextCheckStatus\\"},{\\"Variable\\":\\"$.CheckAlarmStatus.Status\\",\\"StringEquals\\":\\"NO_ALARM\\",\\"Next\\":\\"NoAlarm\\"}],\\"Default\\":\\"BugState\\"},\\"BugState\\":{\\"Type\\":\\"Fail\\"},\\"FailBecauseAlarming\\":{\\"Type\\":\\"Fail\\"},\\"NoAlarm\\":{\\"Type\\":\\"Succeed\\"}}}",
            ],
          ],
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "AlarmMonitorStateMachineRole19D08AAF",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "AlarmMonitorStateMachineRole19D08AAF": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::FindInMap": Array [
                    "ServiceprincipalMap",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    "states",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "AlarmMonitorStateMachineRoleDefaultPolicy7C3EF13E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "AlarmMonitorCheckAlarmStatusFunction38993418",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "AlarmMonitorCheckAlarmStatusFunction38993418",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AlarmMonitorStateMachineRoleDefaultPolicy7C3EF13E",
        "Roles": Array [
          Object {
            "Ref": "AlarmMonitorStateMachineRole19D08AAF",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5bc602ecde93c947efe5899ae355f999986a1acbe610b1c0b9c468d738857555.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestAlarmsAlwaysAlarming69FED027": Object {
      "Properties": Object {
        "ComparisonOperator": "LessThanThreshold",
        "Dimensions": Array [
          Object {
            "Name": "Currency",
            "Value": "USD",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "EstimatedCharges",
        "Namespace": "AWS/Billing",
        "Period": 32400,
        "Statistic": "Maximum",
        "Threshold": 99999999999,
        "TreatMissingData": "breaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarmsCompositeAlwaysAlarmingA7E34597": Object {
      "Properties": Object {
        "AlarmName": "IntegAlarmMonitorTestAlarmsCompositeAlwaysAlarming2A3007AE",
        "AlarmRule": Object {
          "Fn::Join": Array [
            "",
            Array [
              "(ALARM(\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarmsAlwaysAlarming69FED027",
                  "Arn",
                ],
              },
              "\\") OR ALARM(\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarmsNeverAlarmingHopefullyFD878B05",
                  "Arn",
                ],
              },
              "\\"))",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::CompositeAlarm",
    },
    "TestAlarmsCompositeNeverAlarming02CDF5B3": Object {
      "Properties": Object {
        "AlarmName": "IntegAlarmMonitorTestAlarmsCompositeNeverAlarming88E1B001",
        "AlarmRule": Object {
          "Fn::Join": Array [
            "",
            Array [
              "(ALARM(\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestAlarmsNeverAlarmingHopefullyFD878B05",
                  "Arn",
                ],
              },
              "\\"))",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::CompositeAlarm",
    },
    "TestAlarmsNeverAlarmingHopefullyFD878B05": Object {
      "Properties": Object {
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": Array [
          Object {
            "Name": "Currency",
            "Value": "USD",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "EstimatedCharges",
        "Namespace": "AWS/Billing",
        "Period": 32400,
        "Statistic": "Maximum",
        "Threshold": 99999999999,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
